### ISSUE TITLE
FR: use filenames instead of pipes for the pager (where appropriate)

### ISSUE THREAD

**Issue by jyn514** [martinvonz/jj#3013]
**Is your feature request related to a problem? Please describe.**
i am running `jj cat` to see the contents of a file from an old revision. i want to have syntax highlighting, so i set `PAGER='bat --color=always'`. but bat does not know the data coming from stdin is a rust file, so it doesn't generate syntax highlighting; i have to use `PAGER='bat --file-name=cli_util.rs --color=always'` explicitly (which means i can't set this persistently in jj.toml).

**Describe the solution you'd like**
write the file to disk before piping it to the pager.

**Describe alternatives you've considered**
- special-case `bat` and pass it `--file-name` flags
- decide this use case isn't supported and people should just use flags in PAGER explicitly.

**Additional context**

**Comment by joyously** [comment]
Or have an environment variable hold the current file name, to be used as needed?

I think many tools use a file written to /tmp as when an editor allows you to run your compiler on unsaved changes, or the zip file viewer lets you look in a file in the zip.

**Comment by jyn514** [comment]
an environment variable would work, yeah. i'd have to write a little wrapper script that passes it to `bat`, but that's doable.

**Comment by martinvonz** [comment]
> Or have an environment variable hold the current file name, to be used as needed?

Yes, or substitute `$filename` variable like we do for [diff editors](https://martinvonz.github.io/jj/prerelease/config/#editing-diffs). Then the user can set their config like this:

```
ui.pager = ["bat", "--filename", "$filename"]
```

We already allow passing static environment variables (like we do to `less` [by default](https://github.com/martinvonz/jj/blob/785f2f577680d17e69b7d0a194494bfe4aa1b4a2/cli/src/config/misc.toml#L9)). We could allow passing the path that way too, so you could do this:

```
ui.pager = { command = ["my-ager"], env = { MY_PAGER_FILENAME = "$filename" } }
```

But there might not be a need for that, so that's probably just a nice to have.

**Comment by dotcarmen** [comment]
i'm not sure jj will always have a single filename available for the substitution. jyn's use-case certainly does, but jj might invoke the pager for `jj log` (no filename) or `jj diff` (potentially multiple filenames)


i don't think jj will always be able to resolve exactly 1 filename. for example, `jj log` (no filename) or `jj diff` (potentially multiple filenames)

what about an option like `ui.file-formatter`? this would be akin to `ui.diff-formatter`, and all file contents would be processed through this program before being further processed by jj (for example, `jj file annotate -T ...`)

edit: just realized file contents aren't available to the Templating language. i think that's fine, and `jj file annotate` is still a good example of why `ui.file-formatter` would be a good solution
